### PR TITLE
Avoid deprecation warning on explicit eta-expansion of nullary method

### PR DESCRIPTION
**Pull request by som-snytt**
Scala 2.12.4 produces a false deprecation warning for explicit eta-expansion of nullary methods:
```
scala> def f() = 1
scala> f _
<console>:13: warning: Eta-expansion of zero-argument method values is deprecated. Did you intend to write f()?
       f _
```

This PR fixes that.

Nuances https://github.com/scala/scala/commit/c15b9b730e76f2c33399ce4ffc2ffb3c71025531 until method value syntax is deprecated entirely.

Fixes scala/bug#10612

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 1]
```diff
@@ -473,6 +473,13 @@ trait Trees extends api.Trees {
        extends TermTree with TypedApi
   object Typed extends TypedExtractor
 
+  object MethodValue {
+    def unapply(tree: Tree): Option[Tree] = tree match {
+      case Typed(expr, Function(Nil, EmptyTree)) => Some(expr)
+      case _ => None
+    }
+  }
+
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Trees.scala:484]: Let's move `def makeMethodValue` into `MethodValue.apply`.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -5291,12 +5295,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           // however we cannot do `context.withMacrosDisabled`
           // because `expr` might contain nested macro calls (see scala/bug#6673)
           //
-          // Note: apparently `Function(Nil, EmptyTree)` is the secret parser marker
+          // Note: `Function(Nil, EmptyTree)` is the secret parser marker
           // which means trailing underscore -- denoting a method value. See makeMethodValue in TreeBuilder.
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: This comment could be removed now IMO, perhaps replaced with a one line doc comment on `MethodValue` to show the syntactic form it represents.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -909,8 +913,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           checkParamsConvertible(tree, tree.tpe)
 
           // scala/bug#7187 eta-expansion of zero-arg method value is deprecated, switch order of (4.3) and (4.2) in 2.13
-          def isExplicitEtaExpansion = original match {
-            case Typed(_, Function(Nil, EmptyTree)) => true // tree shape for `f _`
+          def isExplicitEtaExpansion = original0 match {
+            case MethodValue(_) => true // tree shape for `f _`
             case _ => false
           }
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: We could move this right up to the first use of `MethodValue.unapply`

```scala
val isExplicitEtaExpansion = original0 != original
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 4]
```diff
@@ -5291,12 +5295,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
           //
-          // Note: apparently `Function(Nil, EmptyTree)` is the secret parser marker
+          // Note: `Function(Nil, EmptyTree)` is the secret parser marker
           // which means trailing underscore -- denoting a method value. See makeMethodValue in TreeBuilder.
-          case Typed(expr, Function(Nil, EmptyTree)) =>
+          case MethodValue(expr) =>
             typed1(suppressMacroExpansion(expr), mode, pt) match {
               case macroDef if treeInfo.isMacroApplication(macroDef) => MacroEtaError(macroDef)
-              case methodValue                                       => typedEta(checkDead(methodValue), expr)
+              case methodValue                                       => typedEta(checkDead(methodValue), tree)
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: So this is crux of the change. Can you describe how the surrounding changes allowed you go back to passing the entire `Typed` tree as `original` here, without incurring the infinite loop I was firefighting in c15b9b730e76f2c33399ce4ffc2ffb3c71025531?

File: src/reflect/scala/reflect/internal/Trees.scala (hunk 1 above)
Comment (som-snytt) [reply src/reflect/scala/reflect/internal/Trees.scala:484]: OK.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 2 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Yes.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 3 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Yes.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 4 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I probably can after I look at the code again.

**Comment by lrytz** [comment]
@retronym this seems to be ready

**Comment by SethTisue** [comment]
@som-snytt Cannily, isn't that where Puff the Magic Dragon lived?

**Comment by som-snytt** [comment]
"I'll have what he's having."

**Comment by SethTisue** [comment]
marking as "prio:blocker" because a lot of people have been asking about it

**Comment by som-snytt** [comment]
I just broke out in a cold sweat.

**Comment by adriaanm** [comment]
I reworked part 2 a bit -- no 💦

**Comment by adriaanm** [comment]
Problem in neg/t10279.scala

**Comment by adriaanm** [comment]
Ok, well, I was a little too hasty here -- sorry! Feel free to push -f your branch to its original state (https://github.com/som-snytt/scala/commit/933de7db5d5339ffc4cd085aab8022496e26e3f8). I was trying to move us away from special casing the treatment of `original` in `adapt`, because it should be a rare case and `adapt` is a core type checker method. I'll take another look on Monday

**Comment by lrytz** [comment]
@som-snytt yet another guy pushing to your branch, sorry... I pair-programmed with @adriaanm on this today, here's what we've come up with. I hope this looks good for 2.12.5.

@SethTisue since you were interested in this issue recently, maybe you'd like to check the test cases to see if everything looks right to you.

File: test/files/neg/t10279.scala [hunk 5]
```diff
@@ -1,13 +1,18 @@
...
+  val t1 = foo(1) _     // error: no implicit string
+  val t2 = foo(1)("") _ // error: _ must follow method
+  val t3 = foo _        // error: no implicit string
+  val t4 = { implicit val s = ""; foo _ } // eta-expansion over the non-implicit parameter list
+  val t4a: Int => String = t4             // ok
+  val t5 = { implicit val s = ""; foo(1) _ }       // compiles as Predef.wrapString(foo(1)(s))
+  val t5a: collection.immutable.WrappedString = t5 // don't ask me why
 
-  val bar = foo(1) _
+  def bar(i: Int)(implicit i: Int): Int = 0
```
Comment (adriaanm) [inline test/files/neg/t10279.scala]: Typo: rename second `i` argument

**Review by adriaanm — CHANGES_REQUESTED** [review]
Very nice! One typo in a test that would be good to fix.

File: test/files/neg/t10279.scala (hunk 5 above)
Comment (lrytz) [reply test/files/neg/t10279.scala]: fixed...

**Review by adriaanm — APPROVED** [review]
💯 great cleanup, exhausti[ve/ng] tests ;-)

**Comment by som-snytt** [comment]
I always wondered what @lrytz means when says "this seems to be ready." I kind of miss `LGTM`.

Also becoming very attached to attachments. They're like post-it notes.

**Comment by SethTisue** [comment]
I'll merge this once Jenkins passes, and then that merge commit will be our RC that I'll send through the gauntlet of the community build

**Comment by hrhino** [comment]
🎉!

**Comment by som-snytt** [comment]
@adriaanm missed opportunity to rename both attachments for tagging `case x @ _` and `f _` to `UnderscoreAttachment`.

**Pull request merged**